### DISCUSSION TITLE
Работа функции Filter.replaceWithSimilar

### DISCUSSION THREAD

**Discussion by ilyanikolaich** [Chimildic/goofy#82] [Принцип работы]
Искренне прошу прощения у нашего Гуру, и у остальных участников обсуждений, но я никак не могу понять логику работу функции replaceWithSimilar():
В описании указано:
>> Заменяет треки на похожие. На одну замену один случайный трек из результатов getRecomTracks. Если замены нет трек удаляется.

Так вот, прошу еще раз прощения за свою непонятливость, но:
replaceWithSimilar( TracksA, TracksB );
Треки массива А заменяются похожими треками из массива В?
Или каждый трек массива А заменяются треками, из массива В, похожими на каждый трек из массива А?
Или каждый трек из массива А заменяется похожими треками массива В?

То есть я не могу осознать - что чему является основой для рекомендаций, и какие треки на основании каких подбираются?

И уж совсем "приземленно" - массив А - 100 треков, массив В - 500 треков, все треки разные - каково оценочное ожидаемое количество треков в А? И наоборот - если в А - 500 треков, а в В - 100 - то чего ожидать от этого?

Полдня сегодня пытался экспериментами получить какие-то результаты, но логики так и не понял.

**Comment by Chimildic** [comment]
> В последних версиях `replaceWithSimilar` имеет другую структуру входных аргументов

Функция `replaceWithSimilar` принимает два массива: где заменять (1) и что заменять (2). 

То есть берем все треки из (1), которые есть в (2). 
Например, в (2) все лайки `Source.getSavedTracks`. Тогда в (1) можно найти эти лайки и на эту же позицию вставить случайный рекомендованный трек, который похож на изначальный лайк. 

В качестве (2) может быть любой массив треков. К примеру, история прослушиваний. Возьмем `микс дня 1`. Мы его послушали наполовину, треки попали в историю. Чтобы снова на них не попасться, можно заменить их на аналог:
```js
Filter.replaceWithSimilar(tracks, RecentTracks.get());
```
Чтобы заменить все:
```js
Filter.replaceWithSimilar(tracks, tracks);
```

Другими словами, (2) это указатель для поиска трека, его позиции в (1). 
Если трек есть в (1), но нет в (2) - его менять не будет.

На один трек - одна замена. Если в (1) - 100 треков, то и максимум будет 100. Но может быть меньше, если для трека не нашлось рекомендаций.
> Удаление трека при отсутствии рекомендаций появилось только вчера. Раньше такой трек просто оставался. Поэтому возможно потребуется обновить код.

**Reply by ilyanikolaich** [reply]
Огромная благодарность за подробное объяснение. Теперь полностью понятна логика работы функции и понятна область ее применения.

**Comment by ilyanikolaich** [comment]
Мне кажется, я наткнулся на ошибку `replaceWithSimilar()`
Исходный плейлист содержит 23 трека.
Мы пробуем заменить в нем все треки, которые слушали - на что-то похожее.
Для проверки эксперимента просто удаляем Историю из Исходного плейлиста - в остатке остается 10 треков, которые не были ранее прослушаны, то есть они должны остаться в модифицированном посредством `replaceWithSimilar` плейлисте.
Однако, полученный плейлист содержит только 13 "новых" треков - которые должны были прийти на замену прослушанным. А те 10, которые должны были остаться в плейлисте - их нет.
```
  let Tracks = Source.getTracks([{ id: '******' }]);
  let History = RecentTracks.get();
  Filter.replaceWithSimilar(Tracks, History);
  Playlist.saveWithReplace({ 
      name: 'TESTING', 
      tracks: Tracks, 
      public: false 
  });
```
Тестировано на разных плейлистах. 
Например попробовать заменить один единственный трек в плейлисте из нескольких треков. И в итоговый плейлисте останется ровно один трек, который отобрался на "замену". То есть можно сформулировать так - в итоговый плейлист не попадают треки, которым не нужно подбирать замену.
Как я понял из разъяснения работы функции, это неверное поведение.
Код, конечно же, обновлен на последний доступный.

**Reply by Chimildic** [reply]
Внес правку

**Reply by Bedmit** [reply]
> Внес правку

Нужно опять обновить библиотеку?
А в каком теперь виде работает функция?

**Reply by Chimildic** [reply]
@Bedmit треки для которых не просили рекомендации останутся на прежнем месте. Треки для которых не нашлось рекомендаций, скорее всего, удалятся.
